src/common/task.c

```diff
@@ -269,47 +281,58 @@ void print_task(task_spec *spec, UT_string *output) {
 
 /* TASK INSTANCES */
 
-struct task_instance_impl {
-  task_iid iid;
-  int32_t state;
+struct task_impl {
+  scheduling_state state;
   node_id node;
   task_spec spec;
 };
 
-task_instance *make_task_instance(task_iid task_iid,
-                                  task_spec *spec,
-                                  int32_t state,
-                                  node_id node) {
-  int64_t size = sizeof(task_instance) - sizeof(task_spec) + task_size(spec);
-  task_instance *result = malloc(size);
+bool node_ids_equal(node_id first_id, node_id second_id) {
+  return UNIQUE_ID_EQ(first_id, second_id) ? true : false;
+}
+
+bool node_id_is_nil(node_id id) {
+  return node_ids_equal(id, NIL_NODE_ID);
+}
+
+task *alloc_task(task_spec *spec, scheduling_state state, node_id node) {
+  int64_t size = sizeof(task) - sizeof(task_spec) + task_spec_size(spec);
```

Comment:
You can't know the `task_size` without allocating the task first. This is to compute the size that `task_size` should return once the task is allocated.